### PR TITLE
CANS-262. Add Multiple Case Numbers to Child Profile

### PR DESCRIPTION
**Pull request by denys-davydov**
## Description
https://osi-cwds.atlassian.net/browse/CANS-262

## Tests
- [x] I used TDD to develop the feature
- [ ] I have included unit tests
- [x] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [x] My code adds no new issues to Code Climate
- [x] I ran all the linters and static analyzers for the project (SonarQube, eslint, rubocop, etc).
- [ ] My code is in a stable state ready to be deployable, but not necessarily complete.
- [ ] I promise on my honor as a CWDS developer to ensure I don't break the build, to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

**Comment by denys-davydov** [comment]
@abelayt Please explain what do you mean by "build locally"?

**Comment by abelayt** [comment]
As per our discussion this morning  and your explanation about building it locally, I'll just take the jenkins result for this review. Thanks!

**Comment by denys-davydov** [comment]
Merging despite the code climate issue, as it doesn't look relevant when the code climate finds duplications for the hibernate annotations on the entity field.

File: src/main/java/gov/ca/cwds/cans/dao/CaseDao.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  public void createOrReplace(final Case aCase, final Person currentUser) {
+    if (aCase.getId() == null) {
+      aCase.setCreatedBy(currentUser);
+      this.create(aCase);
+      return;
+    }
+
+    findByExternalIdOrCreate(aCase, currentUser);
```
Comment (edgibbs) [inline src/main/java/gov/ca/cwds/cans/dao/CaseDao.java]: This method call is never tested.

File: src/main/java/gov/ca/cwds/cans/service/PersonService.java [hunk 2]
```diff
@@ -21,4 +31,38 @@ public PersonService(PersonDao dao) {
...
+
+  @Override
+  public Person create(Person person) {
+    Require.requireNotNullAndNotEmpty(person);
+    initializeCasesForCreate(person.getCases());
+    return super.create(person);
+  }
+
+  private void initializeCasesForCreate(List<Case> cases) {
+    if (CollectionUtils.isEmpty(cases)) {
```
Comment (edgibbs) [inline src/main/java/gov/ca/cwds/cans/service/PersonService.java]: This conditional is never tested, all the test pass whether it exists or not.

File: src/main/java/gov/ca/cwds/cans/service/PersonService.java [hunk 3]
```diff
@@ -21,4 +31,38 @@ public PersonService(PersonDao dao) {
...
+
+  @Override
+  public Person update(Person person) {
+    Require.requireNotNullAndNotEmpty(person);
+    initializeCasesForUpdate(person.getCases());
+    return super.update(person);
+  }
+
+  private void initializeCasesForUpdate(List<Case> cases) {
+    if (CollectionUtils.isEmpty(cases)) {
```
Comment (edgibbs) [inline src/main/java/gov/ca/cwds/cans/service/PersonService.java]: Similar, condition is never evaluated by a test.

File: src/test/java/gov/ca/cwds/cans/rest/resource/PersonResourceTest.java [hunk 4]
```diff
@@ -194,4 +218,129 @@ public void searchPeople_success_whenSearchingForClients() throws IOException {
       assertThat(actualList, hasItem(person));
     }
   }
+
+  @Test
+  public void searchPeople_success_whenSearchingForClients() throws IOException {
```
Comment (edgibbs) [inline src/test/java/gov/ca/cwds/cans/rest/resource/PersonResourceTest.java]: I prefer unit tests with some integrated/functional tests, but this is the current pattern.

File: src/main/java/gov/ca/cwds/cans/service/PersonService.java (hunk 2 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/service/PersonService.java]: @edgibbs According to best practices, we are not testing private methods, only public API. With this said that code still covered by tests indirectly http://sonar.dev.cwds.io:9000/component_measures?id=cwds%3Acans-api&metric=new_lines_to_cover&selected=cwds%3Acans-api%3Asrc%2Fmain%2Fjava%2Fgov%2Fca%2Fcwds%2Fcans%2Fservice%2FPersonService.java

Actual coverage on added code 97% http://sonar.dev.cwds.io:9000/dashboard?id=cwds%3Acans-api

File: src/main/java/gov/ca/cwds/cans/service/PersonService.java (hunk 3 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/service/PersonService.java]: Similar here, private method.

File: src/test/java/gov/ca/cwds/cans/rest/resource/PersonResourceTest.java (hunk 4 above)
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/rest/resource/PersonResourceTest.java]: @edgibbs Fixture based testing for REST APIs provides high enough code coverage (above 80%) while significantly reduces test development time, helps to cover all corner cases using JSON fixtures and fixtures can be used as API documentation. Also the same test can be used as Integration test on the real environment which reduces test development effort.

We are writing fewer tests, spend less time on testing while maintaining high coverage to control regression and reuse tests for unit and integration testing.

File: src/main/java/gov/ca/cwds/cans/dao/CaseDao.java (hunk 1 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/dao/CaseDao.java]: This code is covered with tests according to Sonar http://sonar.dev.cwds.io:9000/component_measures?id=cwds%3Acans-api&metric=new_lines_to_cover&selected=cwds%3Acans-api%3Asrc%2Fmain%2Fjava%2Fgov%2Fca%2Fcwds%2Fcans%2Fdao%2FCaseDao.java

File: src/main/java/gov/ca/cwds/cans/domain/dto/CaseDto.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+import java.time.LocalDateTime;
+import javax.validation.constraints.Pattern;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.experimental.Accessors;
+import org.hibernate.validator.constraints.NotEmpty;
+
+/** @author denys.davydov */
+@Data
+@Accessors(chain = true)
```
[the remark this answers is not in the document]
Comment (denys-davydov) [reply src/main/java/gov/ca/cwds/cans/domain/dto/CaseDto.java]: Yes, there is a convenience reason to do it, because it is nice to build a new object in one line, e.g.:
`new CaseDto().setId(id).setName(name).setDate(date);`

File: src/main/java/gov/ca/cwds/cans/domain/dto/PersonDto.java [hunk 6]
```diff
@@ -27,8 +30,7 @@
 @ToString(callSuper = true)
 @JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
 public class PersonDto extends Dto {
-  @NotNull
-  private PersonRole personRole;
+  @NotNull private PersonRole personRole;
```
[the remark this answers is not in the document]
Comment (denys-davydov) [reply src/main/java/gov/ca/cwds/cans/domain/dto/PersonDto.java]: The code is formatted automatically to comply with Google Java Style Guide that is a best practice for java apps we have on the project https://google.github.io/styleguide/javaguide.html
